base/strings/unicode.jl

```diff
@@ -385,7 +386,10 @@ julia> islowercase('❤')
 false
 ```
 """
-islowercase(c::AbstractChar) = ismalformed(c) ? false : Bool(ccall(:utf8proc_islower, Cint, (UInt32,), UInt32(c)))
+islowercase(c::AbstractChar) = _islowercase(c)
+# identical method body, but with better effects for a Char argument
+@assume_effects :foldable islowercase(c::Char) = _islowercase(c)
+@inline _islowercase(c) = ismalformed(c) ? false : Bool(ccall(:utf8proc_islower, Cint, (UInt32,), UInt32(c)))
```

Comment:
I think it should be enough to mark just the `ccall` as effect-free.

That has the advantage of avoiding this dispatch split and working for `AbstractChar` if `ismalformed` has good effects:
```julia
julia> foo(c::AbstractChar) = Base.ismalformed(c) ? false : UInt32(c)
foo (generic function with 1 method)
julia> Base.infer_effects(foo, (Char,))
(+c,+e,!n,+t,+s,+m,+i)
julia> Base.infer_effects(foo, (AbstractChar,))
(+c,+e,!n,+t,+s,+m,+i)
```

Reply:
I don't really understand this example, as I obtain
```julia
julia> Base.infer_effects(foo, (Char,))
(+c,+e,!n,+t,+s,+m,+u)

julia> Base.infer_effects(foo, (AbstractChar,))
(!c,!e,!n,!t,!s,!m,!u)′
```
In particular, `UInt32(::AbstractChar)` may do anything, so how are the effects being assumed here?

Overall, the suggestion seems reasonable, though.

Reply:
> In particular, UInt32(::AbstractChar) may do anything, so how are the effects being assumed here?

If there are only a handful of methods in the system that match `UInt32(::AbstractChar)` then the compiler may choose to enumerate them all and examine their effects individually. Of course, if a new definition is later provided that matches `UInt32(::AbstractChar)` that means any method that made that assumption

Reply:
It looks like in this case `foo()` is getting good effects on 1.10, but not on current master.

I think that's because inference is actually noticing that the only other definition in the system (for `AnnotatedChar`) is technically recursive, so it may not terminate etc. (although in fact it does): https://github.com/JuliaLang/julia/blob/1221160bc43926efcf0fd7c56a2ab4f0fd062086/base/strings/annotated.jl#L192

Reply:
Thanks, that makes sense